### PR TITLE
Feature Storage and Trust package refactoring (#284)

### PR DESCRIPTION
**Pull request by peppelinux**
* fix: variable name for more redability
* fix: use already existent functions for unsafe parse
* fix: wrong key
* feat: decode_jwt_element refactoring
* chore: removed unused functions
* fix: merged functions
* fix: find_vp_token_key refactoring
* test: added tests for jwt package
* fix: added boundary check
* feat: added as_public_dict
* feat: refactoring
* tests: added federation to the config

* chore: added todo

* feat: implemented methods for trust source

* fix: method name

* chore: module initialization

* feat: implemented TrustSourceData model

* chore: initialized module

* feat: implemented TrustHandlerInterface

* feat: initial implementation of metadata handler

* fix: class name

* feat: initial implementation of DirectTrustJWTHandler

* feat: refactoring of CombinedTrustEvaluator

* Update pyeudiw/jwt/verification.py

* feat: added trust source methods and refactoring

* feat: merged implementations

* chore: moved type

* chore: updated config

* fix: update mongo tests

* fix: fixed model

* chore: removed unused code

* chore: removed unused function

* chore: string interpolation

* feat: dinamyc backend refactoring

* feat: updated inteface

* feat: added dummy Federation handler

* fix: CombinedTrustEvaluator instantiation

* tests: added tests for trust module

* fix: parameter handling

* chore: removed unused files

* feat: db code refactoring

* feat: migrated tests

* fix: minor fixs

* chore: removed unused code

* fix: handle exception properly

* fix: test

* feat: complete refactoring of satosa backend tests

* fix: key handling

* feat: added error class

* chore: fix for test passing

* feat: integrated tests

* feat: merged implementations

* fix: display error message

* fix: possible fix to sd_hash (needs more checks)

* fix: collection name

* fix: minor fixs

* fix: test

* chore: remove unused (to re-integrate)

* feat: helper function

* fix: test

* fix: empty behavior

* Update pyeudiw/sd_jwt/sd_jwt.py

* Update pyeudiw/openid4vp/vp_sd_jwt_vc.py

* Update pyeudiw/storage/db_engine.py

* Update pyeudiw/tests/trust/test_dynamic.py

* fix: naming

* chore; added integration test in CI

* fix: path

* fix: escaping

* fix: added dependency

* Apply suggestions from code review

* feat: upodated and refactored trust package

---------

### CODE REVIEW

File: pyeudiw/jwt/parse.py [hunk 1]
```diff
@@ -24,23 +24,18 @@ def parse(jws: str) -> 'DecodedJwt':
...
-    payload = {}
+    
     try:
-        head = _unsafe_decode_part(b64header)
-        payload = _unsafe_decode_part(b64payload)
+        head = decode_jwt_header(token)
+        payload = decode_jwt_payload(token)
+        signature = token.split(".")[2]
     except Exception as e:
         raise ValueError(f"unable to decode JWS part: {e}")
```
Comment (peppelinux) [inline pyeudiw/jwt/parse.py:40]:
```suggestion
        raise ValueError(f"unable to decode JWT: {e}")
```

File: pyeudiw/storage/mongo_storage.py [hunk 2]
```diff
@@ -209,48 +212,69 @@ def update_response_object(self, nonce: str, state: str, internal_response: dict
...
+        )
+    
+    def has_trust_source(self, entity_id: str) -> bool:
+        return self._has_db_entity(
+            self.storage_conf["db_trust_sources_collection"], 
+            entity_id
+        )
 
-    def _add_entry(
+    def _upsert_entry(
```
Comment (peppelinux) [inline pyeudiw/storage/mongo_storage.py:258]: this should be implemented also in the base_storage

File: pyeudiw/storage/mongo_storage.py [hunk 3]
```diff
@@ -209,48 +212,69 @@ def update_response_object(self, nonce: str, state: str, internal_response: dict
...
+
+        document_status = db_collection.update_one(
+            {key_label: data[key_label]},
+            {"$set": data},
+            upsert=True
+        )
+
+        if not document_status.acknowledged:
+            raise StorageEntryUpdateFailed(
+                "Trust Anchor matched count is ZERO"
```
Comment (peppelinux) [inline pyeudiw/storage/mongo_storage.py:274]: please use a more eloquent error message

File: pyeudiw/storage/mongo_storage.py [hunk 4]
```diff
@@ -281,6 +308,7 @@ def _update_anchor_metadata(self, entity: dict, attestation: list[str], exp: dat
 
         entity[trust_name] = trust_entity
 
+
```
Comment (peppelinux) [inline pyeudiw/storage/mongo_storage.py:311]:
```suggestion
```

File: pyeudiw/trust/handler/direct_trust_sd_jwt_vc.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        "timeout": os.getenv("PYEUDIW_HTTPC_TIMEOUT", 6)
+    }
+}
+
+class DirectTrustSdJwtVc(TrustHandlerInterface, BaseLogger):
+    def __init__(
+            self, 
+            httpc_params: dict = DEFAULT_DIRECT_TRUST_SD_JWC_VC_PARAMS, 
+            jwk_endpoint: str = DEAFAULT_JWK_ENDPOINT,
+            metadata_endpoint: str = DEAFAULT_METADATA_ENDPOINT,
```
Comment (peppelinux) [inline pyeudiw/trust/handler/direct_trust_sd_jwt_vc.py:29]:
```suggestion
            jwk_endpoint: str = DEFAULT_SDJWTVC_METADATA_ENDPOINT,
            metadata_endpoint: str = DEFAULT_OPENID4VCI_METADATA_ENDPOINT,
```

File: pyeudiw/trust/handler/direct_trust_sd_jwt_vc.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+        baseurl = urlparse(issuer_id)
+        well_known_path = well_known_path_component + baseurl.path
+        well_known_url: str = ParseResult(baseurl.scheme, baseurl.netloc, well_known_path, baseurl.params, baseurl.query, baseurl.fragment).geturl()
+        return well_known_url
+
+    def build_issuer_metadata_endpoint(issuer: str, metadata_path_component: str) -> str:
+        issuer_normalized = issuer if issuer[-1] != '/' else issuer[:-1]
+        return issuer_normalized + metadata_path_component
+    
+
```
Comment (peppelinux) [inline pyeudiw/trust/handler/direct_trust_sd_jwt_vc.py:85]:
```suggestion
```

**Comment by Zicchio** [comment]
I think that this function
https://github.com/italia/eudi-wallet-it-python/blob/5c3c95ea1918aedbea140bc675bfedc9732bac97/pyeudiw/jwt/parse.py#L49-L57

should be clearly delegated to the trust model implementation(s). Not only the key identifier might not exist in some trust model, but the methods stubs `get_public_key_from_trust_chain` and `get_public_key_from_x509_chain` clearly reference functionalities that only trust implementations can provide.
This is not a minor detail! The lack of this functionality, in the current code, actually __BLOCKS ANY token verification method to ever work in any model different that direct trust__.

We tried to provide a one and only trust evaluator that is completely agnostic to the token representation, but IMO this cannot work. This is also hinted in this exception, which is basically impossible to overcome as the given function is completely detached from any object ably to parse or vallidate a trust source (i.e. trust proof/attestation) 
https://github.com/italia/eudi-wallet-it-python/blob/7330f41bcfe7885cc07f38b96ba5a6aab7478a65/pyeudiw/tools/jwk_handling.py#L28

Moreover, only a the trust model can know which parts of a token header can be treated as trust attestation or proof that a public key is valid, hence a I am keen to conclude that the **trust model SHOULD be able to parse token headers** . For example, only a federation trust model can know the fact that it can find relevant information in, say, the claim `trust_chain` of a token header.

I think that it is best to catch two birds with one stone. The trust aware interface (TrustHandler, TrustEvaluator or whatever it might be called) might implement a functionality similar to the following.

```
class TrustEvaluator:
....
    def obtain_signing_key(self, issuer: str, token_header: dict) -> dict:
    """Find a public signing key of the issuer that can be uniquely referenced and trusted
    based on the content of the given token header it might verify.
    """
```

we could also work a more abtract and modular description that emphasize the trust attestation / proof / source, such as
```
....
    def obtain_attestation_from_token_header(self, issuer: str, token_header: dict) -> dict<trust_attestation>:
        ...
    def obtain_public_key(self, issuer: str, trust_attestation: dict) -> dict:
        ...
```

but the core idea is the same: the trust implementation should be able to parse and process token header in order to yield valid cryptographic material for _that_ token.

I don't love the conclusion, as it basically means that we might struggle in extending the project if/when we need to process verifiable credentials that are not in the form of a sd-jwt, but I honeslty fail to see a way to overome these issues.

@peppelinux waiting an input on the topic by you.

**Comment by peppelinux** [comment]
Closed in favor of https://github.com/italia/eudi-wallet-it-python/pull/315

**Pull request closed without merging**